### PR TITLE
KFSPTS-22326 Add support for Request V4 API

### PR DESCRIPTION
**Pull request by cah292**
Please do not merge this yet until I add a comment stating otherwise. I'm still working with Tami to get this tested locally and to potentially fix some problems with the Concur test workflow.

There are PRs for this change in both cu-kfs and nonprod-sql. Please make sure both are good to go before merging.

This PR updates the new Concur Event Notification V2 batch job so that it can process Concur travel requests via the Request V4 API. Note that the changes below are only adding the ability to read and validate the requests; the part that updates the report's status in Concur will be implemented in a future user story.

Using the get-requests part of the V4 API to fit our needs is a bit complicated. We're looking for requests in "Pending External Validation" status, but such requests don't show up in queries that filter according to the "TOAPPROVE" or "PENDING" views. They do show up in the "APPROVED" view, and that view can be used to retrieve all the relevant requests or just the ones approved by specific test users. Using a mix of any-approver and specific-approver searching, we can use the API to find the pending requests and filter them based on whether or not the current environment is a Production one. If you think there's a better way to do this, please let me know.

The unit tests build upon some helper web-service-related classes that are relied upon or introduced by prior user stories. This PR performs some extra adjustments to them, to make some of the setup easier and to make the functionality compatible with JUnit 5. (The relied-upon LocalServerTestBase class was originally written for legacy JUnit.)

The main unit test class also does some extra zone-related conversion of the test dates, to keep the "local" time zone consistent when parsing date strings that are not specifically in UTC time. Hopefully this will keep the unit test behavior consistent if the tests get executed in a different time zone.

### CODE REVIEW

**Comment by cah292** [comment]
retest this please

**Comment by jhulslander** [comment]
retest this please

**Comment by cah292** [comment]
With the current setup of this new process, the last-processed-date parameter won't get updated unless the processing of the travel requests completes without uncaught exceptions being thrown. Thus, the next run should reprocess the item that failed last time, and the other successful requests should theoretically have moved to "Approved" status in Concur so they won't get processed twice.

There's still more work that could be done to improve the error handling for the Concur API calls, so that operation errors on a single request won't force a short-circuit of the iteration over the other requests. However, given the complexity of this PR, I would like to defer the error handling cleanup to the follow-up user story that implements the workflow-updating calls.

File: src/test/java/edu/cornell/kfs/concur/batch/service/impl/MockConcurRequestV4Server.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    public MockConcurRequestV4Server(String baseRequestV4Url, RequestV4DetailFixture... requestsToAdd) {
+        this.travelRequests = new ConcurrentHashMap<>();
+        this.baseRequestV4Url = baseRequestV4Url;
+        addTravelRequests(requestsToAdd);
+    }
+
+    @Override
+    public void close() throws IOException {
+        if (travelRequests != null) {
+            travelRequests.clear();
```
Comment (jhulslander) [inline src/test/java/edu/cornell/kfs/concur/batch/service/impl/MockConcurRequestV4Server.java:55]: is this needed?  Shouldn't setting travelRequests to null clear everything up?

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java [hunk 2]
```diff
@@ -0,0 +1,432 @@
...
+    protected ConcurBatchUtilityService concurBatchUtilityService;
+    protected ConcurEventNotificationV2WebserviceService concurEventNotificationV2WebserviceService;
+    protected ConcurAccountValidationService concurAccountValidationService;
+    protected ConfigurationService configurationService;
+    protected DateTimeService dateTimeService;
+
+    @Override
+    public List<ConcurEventNotificationProcessingResultsDTO> processTravelRequests(String accessToken) {
+        LOG.info("processTravelRequests, Starting processing of travel requests");
+        RequestV4QuerySettings querySettings = new RequestV4QuerySettings(
```
Comment (jhulslander) [inline src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java:73]: I am finding this code hard to follow.  Perhaps it could use some refactoring to help with clarity.  My expectation for the logic flow is below.  Is my expectation too simplistic 

1. Get the travel requests in pending external validation based on the time timeframe logic you have figured out
2. If non-prod, filter the list for only test users, if prod, filter the list for only users that aren't test users
3. Loop through the list validating the accounting details
4. Accumulate ResultsDTO objects as you
5. Leave a holding place to update the status in concur on a future user story

**Review by jhulslander — CHANGES_REQUESTED** [review]
I have done some code review, but got stuck on ConcurRequestV4ServiceImpl .... I am wondering if it could use some refactoring to help clarity.  Also, I noted my expectation for what I thought the flow should be.  I am not sure if my expectation is too simplistic or if ConcurRequestV4ServiceImpl could be simplified, or a combination.

File: src/main/resources/edu/cornell/kfs/concur/cu-concur-resources.properties [hunk 3]
```diff
@@ -0,0 +1,2 @@
+message.concur.requestv4.listing=Travel Request Listing (page {0})
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/main/resources/edu/cornell/kfs/concur/cu-concur-resources.properties:1]: I didn't find any user stories for migrating those properties. Given that multiple modules have properties within this file, I will probably create an initial analysis user story for determining which modules have data in CU Application Resources, and then we can use separate user stories or sub-tasks to split out the properties on a per-module basis.

File: src/test/java/edu/cornell/kfs/concur/batch/service/impl/MockConcurRequestV4Server.java (hunk 1 above)
Comment (cah292) [reply src/test/java/edu/cornell/kfs/concur/batch/service/impl/MockConcurRequestV4Server.java:55]: I agree that this clear() operation is not needed for the unit testing. I'll remove it.

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java (hunk 2 above)
Comment (cah292) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java:73]: The steps you listed are the general logic that the current code is following. However, the filter-for-users step is a little more complex than merely filtering based on the approval user:

* When a request is Pending External Validation, the JSON doesn't show an active approver, and the prior approvers are not shown either.
* To determine which requests were approved by a test user, I add the test user's ID to the query parameters, and record the UUIDs of the resulting requests.
* To determine which requests should actually be processed, I run the query without specifying a user to filter on, and then I filter the results further based on whether or not they appear in the set of test-user-approved results.

I could simplify the is-test-request logic if there was some other easy-to-analyze field I could check. I could try checking the request's "owner" instead of the approver, as long as we can expect the test requests to only be created by test users. (I'm not familiar with Concur's user account setup, so I wasn't certain whether someone ever used Production accounts to create a test request.) If limiting based on owner is possible, then I could update the code to check for a known set of test users (Cheri, etc.).

Comment (jhulslander) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java:73]: I like the idea of filtering by owner.  I expect that Tami is only going to test with Cher Knapp, or maybe a few other small subset of people.  I know Cheri will never have a production travel request.  We should check with Tami.

**Comment by cah292** [comment]
I've committed some follow-up changes to simplify the is-test-request logic, and to also perform some miscellaneous cleanup and refactoring. I still need to create the appropriate user stories for the follow-up work, though (or at least add the details to a sub-task for creating such stories).

This is ready for a second round of review, but it is not ready to be merged yet. I still need Tami to create some more test requests, and I still need Tami to confirm what all the UUIDs are for the various test users. As an alternative to the latter, I could update the code to find and retrieve the appropriate UUIDs at runtime (so that they're not stored in one long concatenated parameter); however, Concur would have to grant us access to their Identity V4 API.

**Review by jhulslander — CHANGES_REQUESTED** [review]
I had a few questions, and a couple suggestions that may simplify this implementation

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+                .flatMap(List::stream)
+                .filter(operation -> StringUtils.equalsIgnoreCase(ConcurApiOperations.NEXT, operation.getName()))
+                .map(ConcurRequestV4OperationDTO::getHref)
+                .peek(this::verifyQueryUrlForNextResultsPageIsValid)
+                .findFirst()
+                .orElse(KFSConstants.EMPTY_STRING);
+    }
+
+    protected void verifyQueryUrlForNextResultsPageIsValid(String queryUrl) {
+        if (!ConcurUtils.validateFormatAndPrefixOfParameterizedUrl(queryUrl, getRequestV4Endpoint())) {
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java:125]: The code in this PR takes the next-page URL returned by Concur's web service and calls it verbatim. I added this minimal validation to make sure we didn't receive a malicious URL for some reason. (Such a situation is unlikely, but I'm not comfortable with blindly trusting the URL returned from an external system.)

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,401 @@
...
+        Stream.Builder<Stream<T>> subResults = Stream.builder();
+        Stream<T> subResult;
+        String initialQueryUrl = buildInitialRequestQueryUrl(querySettings);
+        String currentQueryUrl = initialQueryUrl;
+        int page = 1;
+        ConcurRequestV4ListingDTO requestListing;
+        
+        do {
+            requestListing = findPendingTravelRequests(querySettings.accessToken, page, currentQueryUrl);
+            subResult = requestListingProcessor.apply(querySettings, requestListing);
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java:102]: No, the "requestListingProcessor" is what takes the intermediate Request Listings, and iterates over their Requests to validate their accounting info and return the results. I had left this as a BiFunction since it simplified my setup of the unit tests that simply retrieve and validate the Request Listing objects. I agree that its setup is still a little confusing, though, so I'll see if I can simplify this part further and adjust the unit tests accordingly.

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,401 @@
...
+                .filter(operation -> StringUtils.equalsIgnoreCase(ConcurApiOperations.NEXT, operation.getName()))
+                .map(ConcurRequestV4OperationDTO::getHref)
+                .peek(this::verifyQueryUrlForNextResultsPageIsValid)
+                .findFirst()
+                .orElse(KFSConstants.EMPTY_STRING);
+    }
+
+    protected void verifyQueryUrlForNextResultsPageIsValid(String queryUrl) {
+        if (!ConcurUtils.validateFormatAndPrefixOfParameterizedUrl(queryUrl, getRequestV4Endpoint())) {
+            throw new RuntimeException("Query URL for next page of search results is malformed");
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java:126]: Good point, I agree that the bad URL needs to be logged somewhere. I'll add it to the exception and/or a separate log statement.

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+        }
+    }
+
+    protected ConcurRequestV4ListingDTO findPendingTravelRequests(String accessToken, int page, String queryUrl) {
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java:130]: Ok, I'll update this to use the "get" convention from the Expense V3 service.

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,401 @@
...
+            String accessToken, String requestUuid, ConcurEventNotificationProcessingResultsDTO resultsDTO) {
+        boolean isValid =
+                (resultsDTO.getProcessingResults() == ConcurEventNotificationVersion2ProcessingResults.validAccounts);
+        LOG.info("updateRequestStatusInConcur, Will notify Concur that Request " + requestUuid
+                + " had an overall validation of " + isValid);
+        
+        LOG.info("updateRequestStatusInConcur, Update of the Request's status in Concur has not been implemented yet");
+    }
+
+    protected ConcurRequestV4ReportDTO findTravelRequest(String accessToken, String requestUuid) {
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java:238]: I'll update this to the "get" convention.

File: src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,401 @@
...
+    protected String getTestUserNameFromEntry(String keyValuePair) {
+        return StringUtils.substringBefore(keyValuePair, CUKFSConstants.EQUALS_SIGN);
+    }
+
+    protected String getTestUserUuidFromEntry(String keyValuePair) {
+        return StringUtils.substringAfter(keyValuePair, CUKFSConstants.EQUALS_SIGN);
+    }
+
+    protected String getLastModifiedFromDateInUTCFormat() {
+        String dateString = getNonBlankConcurParameterValue(ConcurParameterConstants.REQUEST_V4_QUERY_FROM_DATE);
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/main/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImpl.java:307]: The Travel Request V4 API does not appear to have a way to specifically query for requests in Pending External Validation status, so I have to retrieve a broader set of requests that includes some which do not need processing. The last-modified-date strategy helps keep the number of non-relevant requests to a minimum; however, given that the volume of travel requests over a small timeframe is likely not very large, doing a look-back-X-days strategy instead should still keep the load of non-relevant requests low enough. I'll look into adjusting the code accordingly to use the latter.

File: src/test/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImplTest.java [hunk 10]
```diff
@@ -0,0 +1,764 @@
...
+        requestV4Service.setDateTimeService(buildSpiedDateTimeService(testDateTimeService));
+        requestV4Service.setConcurAccountValidationService(buildMockConcurAccountValidationService());
+    }
+
+    @SuppressWarnings("deprecation")
+    @AfterEach
+    void tearDown() throws Exception {
+        IOUtils.closeQuietly(mockConcurEndpoint);
+        IOUtils.closeQuietly(mockHttpServer);
+        IOUtils.closeQuietly(mockConcurBackendServer);
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/test/java/edu/cornell/kfs/concur/batch/service/impl/ConcurRequestV4ServiceImplTest.java:143]: A few of these test classes don't have resources or other handlers being left open, so I could potentially exclude their close() logic and let the Java GC clean them up instead. However, the "mockHttpServer" actively runs a local HTTP server, so that will still need to be closed or shut down accordingly. I may rename that variable to keep the intent a little clearer.

File: src/test/java/edu/cornell/kfs/concur/batch/service/impl/MockConcurRequestV4Server.java [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+        Date modifiedAfter = getExistingDateParameter(queryParameters, ConcurApiParameters.MODIFIED_AFTER);
+        Date modifiedBefore = getExistingDateParameter(queryParameters, ConcurApiParameters.MODIFIED_BEFORE);
+        String sortField = getExistingParameter(queryParameters, ConcurApiParameters.SORT_FIELD);
+        String sortOrder = getExistingParameter(queryParameters, ConcurApiParameters.SORT_ORDER);
+        
+        Comparator<ConcurRequestV4ListItemDTO> requestSorter = Comparator.comparing(
+                ConcurRequestV4ListItemDTO::getStartDate, Comparator.reverseOrder());
+        
+        if (!StringUtils.equalsIgnoreCase(RequestV4Views.APPROVED, view)) {
+            throw new IllegalArgumentException(
```
[the remark this answers is not in the document]
Comment (cah292) [reply src/test/java/edu/cornell/kfs/concur/batch/service/impl/MockConcurRequestV4Server.java:90]: This mock server is still trying to simulate a portion of the Request V4 search functionality. Certain types of invalid inputs (like an out-of-bound index or a malformed date) could cause the simulated search to fail without warning, so having this particular validation in place will make it easier to spot potential errors in our Production code or in our JAX-RS Client setup.

**Comment by cah292** [comment]
Due to the age of this branch, I've created a fresher feature branch for implementing the remaining changes. Further review and discussion should be done on PR #1284 instead.
